azurerm/internal/services/securitycenter/resource_arm_security_center_automation.go

```diff
@@ -0,0 +1,567 @@
+			}
+			autoAction = security.AutomationActionLogicApp{
+				LogicAppResourceID: &resourceID,
+				URI:                &triggerURL,
+				ActionType:         security.ActionTypeLogicApp,
+			}
+
+		// Handle LogAnalytics action type
+		case typeLogAnalytics:
+			autoAction = security.AutomationActionWorkspace{
+				WorkspaceResourceID: &resourceID,
+				ActionType:          security.ActionTypeWorkspace,
+			}
+
+		// Handle EventHub action type
+		case typeEventHub:
+			var connString string
+			if connString, ok = actionMap["connection_string"].(string); !ok || connString == "" {
+				return nil, fmt.Errorf("Security Center automation, connection_string is required for EventHub action")
+			}
+			autoAction = security.AutomationActionEventHub{
+				EventHubResourceID: &resourceID,
+				ConnectionString:   &connString,
+				ActionType:         security.ActionTypeEventHub,
+			}
+		default:
+			continue
+		}
+		output = append(output, autoAction)
+	}
+
+	return &output, nil
+}
+
+func flattenSources(sources *[]security.AutomationSource) []map[string]interface{} {
+	if sources == nil {
+		return make([]map[string]interface{}, 0)
+	}
+
+	resultSlice := make([]map[string]interface{}, 0)
+	for _, source := range *sources {
+		ruleSetSlice := make([]interface{}, 0)
+
+		// RuleSets is an optional field need check for nil
+		if source.RuleSets != nil {
+			for _, ruleSet := range *source.RuleSets {
+				ruleSlice := make([]map[string]string, 0)
+
+				for _, rule := range *ruleSet.Rules {
+					ruleMap := map[string]string{
+						"property_path":  *rule.PropertyJPath,
+						"expected_value": *rule.ExpectedValue,
+						"operator":       string(rule.Operator),
+						"property_type":  string(rule.PropertyType),
+					}
+					ruleSlice = append(ruleSlice, ruleMap)
+				}
+
+				ruleSetMap := map[string]interface{}{
+					"rule": ruleSlice,
+				}
+				ruleSetSlice = append(ruleSetSlice, ruleSetMap)
+			}
+		}
+
+		sourceMap := map[string]interface{}{
+			"event_source": source.EventSource,
+			"rule_set":     ruleSetSlice,
+		}
+		resultSlice = append(resultSlice, sourceMap)
+	}
+
+	return resultSlice
+}
+
+func flattenScopes(scopes *[]security.AutomationScope) []string {
+	if scopes == nil {
+		return []string{}
+	}
+
+	resultSlice := make([]string, 0)
+	for _, scope := range *scopes {
+		if scope.ScopePath == nil {
+			continue
+		}
+
+		resultSlice = append(resultSlice, *scope.ScopePath)
+	}
+
+	return resultSlice
+}
+
+func flattenActions(actions *[]security.BasicAutomationAction, d *schema.ResourceData) []map[string]string {
+	if actions == nil {
+		return []map[string]string{}
+	}
+
+	// Get existing schema data for actions
+	schemaDataActions, schemaDataActionsOk := d.GetOk("action")
+
+	resultSlice := make([]map[string]string, 0)
+
+	for i, action := range *actions {
+		// Use type assertion to discover the underlying action
+		actionLogicApp, isLogicApp := action.(security.AutomationActionLogicApp)
+		if isLogicApp {
```

Comment:
This was my original approach I wanted to use a switch, however trying to convert the interface like this `action.(security.AutomationAction).ActionType` results in a panic
```
panic: interface conversion: security.BasicAutomationAction is security.AutomationActionWorkspace, not security.AutomationAction
```
I don't know why, it _should_ work, but it doesn't. I think it's a bug in the Go SDK (and there's a LOT of them). I lost a lot of time trying various things until I settled on the "multiple if" approach you see in the code

Reply:
fwiw the SDK has an `As[XXX]` method (returning `{type}, ok`) on these "discriminator" types - which should avoid the panicking problem - but they're not ideal for other reasons

Reply:
I also tried that, for whatever reason `action.AsAutomationAction()` always returns false in the `ok` 

But if you call the `As[xxx]` of the more concrete types, e.g. `action.AsAutomationActionLogicApp()` it will work, but then my code flow ends up exactly the same as using the go type assertion (which I prefer over these `As[xxx]` methods, it's a little more obvious)